text/0648-aspects-priority-ordering.md

```diff
@@ -0,0 +1,346 @@
+
+# Priority-Ordered Aspect Invocation
+
+* **Original Author(s):**: @sumupitchayan
+* **Tracking Issue**: #648
+* **API Bar Raiser**: @rix0rrr @mrgrain
+
+Redesigning Aspect invocation in CDK by allowing users to specify the order in which Aspects are applied.
+
+## Working Backwards
+
+### CHANGELOG
+
+#### feat(core): Priority-Ordered Aspect Invocation
+
+* Introduced a priority-based ordering system for aspects in the CDK to allow users to control the order in which aspects are applied across constructs.
+* Added default priority ranges to assist with common use cases (e.g., mutating aspects, readonly aspects) and to improve the execution flow of aspects.
+
+### README
+
+Aspects is a feature in CDK that allows you to apply operations or transformations across all constructs in a construct tree. Common use cases include
+tagging resources, enforcing encryption on S3 Buckets, or applying specific security or compliance rules to all resources in a stack.
+
+Conceptually, there are two types of Aspects:
+
+* Read-only aspects scan the construct tree but do not make changes to the tree. Common use cases of read-only aspects include performing validations
+(for example, enforcing that all S3 Buckets have versioning enabled) and logging (for example, collecting information about all deployed resources for
+audits or compliance).
+* Mutating aspects either (1.) add new nodes or (2.) mutate existing nodes of the tree in-place. One commonly used mutating Aspect is adding Tags to
+resources. An example of an Aspect that adds a node is one that automatically adds a security group to every EC2 instance in the construct tree if no
+default is specified.
+
+Users can ensure Aspects are applied in a predictable and controlled order by using the optional priority parameter when applying an Aspect. Priority
+values must be non-negative integers, where a higher number means the Aspect will be applied later, and a lower number means it will be applied sooner.
+
+CDK provides standard priority values for mutating and readonly aspects to help ensure consistency across different construct libraries:
+
+```ts
+const MUTATING_PRIORITY = 100;
+const READONLY_PRIORITY = 300;
```

Comment:
If we are going with 3 digits, let's at least use the range better
```suggestion
const MUTATING_PRIORITY = 200;
const READONLY_PRIORITY = 800;
```

Reply:
I don't have super strong opinions here. My suggestion is to have:

```console
const MUTATING_PRIORITY = 200;
const READONLY_PRIORITY = 1000;
const DEFAULT_PRIORITY = 600;
```

This will help people get into the habit that `READONLY` is 4 digits, and `MUTATING` is 3 digits, and thus mutating will run before read-only. 

Default values is than jus the average, because why not.